File: app/src/main/java/com/twt/zhihu/zhihudaily/Presenter/ListAdapter.java
package com.twt.zhihu.zhihudaily.Presenter;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.twt.zhihu.zhihudaily.Model.MainBean;
import com.twt.zhihu.zhihudaily.R;
import com.twt.zhihu.zhihudaily.View.ContentActivity;

import java.util.ArrayList;
import java.util.List;

public class ListAdapter extends RecyclerView.Adapter<ListAdapter.ViewHolder> {
    private Context context;
    public List<MainBean> mainBeanList;
    private int times = 0;
    private static final int HEADER = 0;
    private static final int LIST = 1;
    private static final int ITEM_DATE = 0;

    public ListAdapter(Context context, MainBean latestBean) {
        this.context = context;
        mainBeanList = new ArrayList<>();
        mainBeanList.add(latestBean);
    }

    public void addData(MainBean latestBean) {
        mainBeanList.add(latestBean);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == HEADER) {
            BannerAdapter bannerAdapter;
            bannerAdapter = (BannerAdapter) LayoutInflater.from(context).inflate(R.layout.item_banner, parent, false);
            return new ViewHolder(bannerAdapter);
        } else {
            View view = LayoutInflater.from(context).inflate(R.layout.item_layout, parent, false);
            return new ViewHolder(view);
        }
    }

    @Override
    public int getItemViewType(int position) {
        return position == HEADER ? HEADER : LIST;         //判断是否为头部
    }


    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
        int realPosition = position - 1;
        if (position == HEADER) {
            ((BannerAdapter) holder.view).setList(mainBeanList.get(0).top_stories);         //头部为ViewPager
        } else {
            int group = 0;
            int length = 0;
            for (MainBean latestBean : mainBeanList) {
                if (realPosition >= latestBean.stories.size()) {
                    if (realPosition != latestBean.stories.size() || length <= latestBean.stories.size()) {
                        realPosition -= latestBean.stories.size();
                        group++;
                    }
                }
                length = latestBean.stories.size();
            }

            if(realPosition >= mainBeanList.get(group).stories.size()){
                realPosition -= mainBeanList.get(group).stories.size();
                group++;
            }                                               //解决前后两天列表长度不一而导致的group错乱的bug

            final int bean = group;
            final int beanPosition = realPosition;
            holder.textView.setText(mainBeanList.get(bean).stories.get(beanPosition).title);
            Picasso.with(context)
                    .load(mainBeanList.get(bean).stories.get(beanPosition).images.get(0))
                    .fit()
                    .centerCrop()
                    .into(holder.imageView);
            holder.cardView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(context, ContentActivity.class);
                    Bundle bundle = new Bundle();
                    bundle.putString("id", String.valueOf(mainBeanList.get(bean).stories.get(beanPosition).id));
                    intent.putExtras(bundle);
                    context.startActivity(intent);
                }
            });

            if (beanPosition == ITEM_DATE) {
                holder.item_date.setVisibility(View.VISIBLE);
                if (bean != 0) {

                    holder.item_date.setText(mainBeanList.get(bean).date);
                } else {
                    holder.item_date.setText("今日热闻");
                }                                               //日期栏
            } else {
                holder.item_date.setVisibility(View.GONE);
            }                                                   //其余时间隐藏


        }
    }

    @Override
    public int getItemCount() {
        int count = 0;
        for (MainBean latestBean : mainBeanList) {
            count += latestBean.stories.size();
        }
        return count + 1;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView textView;
        ImageView imageView;
        TextView item_date;
        View view;
        View cardView;

        ViewHolder(View itemView) {
            super(itemView);
            view = itemView;
            item_date = itemView.findViewById(R.id.item_date);
            cardView = itemView.findViewById(R.id.cardview);
            textView = itemView.findViewById(R.id.tv);
            imageView = itemView.findViewById(R.id.iv);
        }

    }

}

File: app/src/main/java/com/twt/zhihu/zhihudaily/presenter/ListAdapter.kt
package com.twt.zhihu.zhihudaily.presenter

import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.squareup.picasso.Picasso
import com.twt.zhihu.zhihudaily.R
import com.twt.zhihu.zhihudaily.model.MainBean

import com.twt.zhihu.zhihudaily.View.ContentActivity

import java.util.ArrayList

class ListAdapter(private val context: Context, latestBean: MainBean) : RecyclerView.Adapter<ListAdapter.ViewHolder>() {
    var mainBeanList: MutableList<MainBean> = ArrayList()

    init {
        mainBeanList.add(latestBean)
    }

    fun addData(latestBean: MainBean) {
        mainBeanList.add(latestBean)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return if (viewType == HEADER) {
            val bannerAdapter: BannerAdapter = LayoutInflater.from(context).inflate(R.layout.item_banner, parent, false) as BannerAdapter
            ViewHolder(bannerAdapter)
        } else {
            val view = LayoutInflater.from(context).inflate(R.layout.item_layout, parent, false)
            ViewHolder(view)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == HEADER) HEADER else LIST         //判断是否为头部
    }


    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        var realPosition = position - 1
        if (position == HEADER) {
            (holder.view as BannerAdapter).setList(this.mainBeanList[0].top_stories!!)
            //头部为ViewPager
        } else {
            var group = 0
            var length = 0
            for (latestBean in mainBeanList) {
                if (realPosition >= latestBean.stories.size) {
                    if (realPosition != latestBean.stories.size || length <= latestBean.stories.size) {
                        realPosition -= latestBean.stories.size
                        group++
                    }
                }
                length = latestBean.stories.size
            }

            if (realPosition >= mainBeanList[group].stories.size) {
                realPosition -= mainBeanList[group].stories.size
                group++
            }                                               //解决前后两天列表长度不一而导致的group错乱的bug

            val bean = group
            val beanPosition = realPosition
            holder.textView?.text = mainBeanList[bean].stories[beanPosition].title
            Picasso.with(context)
                    .load(mainBeanList[bean].stories[beanPosition].images!![0])
                    .fit()
                    .centerCrop()
                    .into(holder.imageView)
            holder.cardView?.setOnClickListener {
                val intent = Intent(context, ContentActivity::class.java)
                val bundle = Bundle()
                bundle.putString("id", mainBeanList[bean].stories[beanPosition].id.toString())
                intent.putExtras(bundle)
                context.startActivity(intent)
            }

            if (beanPosition == ITEM_DATE) {
                holder.item_date?.visibility = View.VISIBLE
                if (bean != 0) {

                    holder.item_date?.text = mainBeanList[bean].date
                } else {
                    holder.item_date?.text = "今日热闻"
                }                                               //日期栏
            } else {
                holder.item_date?.visibility = View.GONE
            }                                                   //其余时间隐藏


        }
    }

    override fun getItemCount(): Int {
        var count = 0
        for (latestBean in mainBeanList) {
            count += latestBean.stories.size
        }
        return count + 1
    }

    class ViewHolder(var view: View) : RecyclerView.ViewHolder(view) {
        var textView: TextView? = view.findViewById(R.id.tv)
        var imageView: ImageView? = view.findViewById(R.id.iv)
        var item_date: TextView? = view.findViewById(R.id.item_date)
        var cardView: View? = view.findViewById(R.id.cardview)

    }

    companion object {
        private const val HEADER = 0
        private const val LIST = 1
        private const val ITEM_DATE = 0
    }

}
